File: day2.py
import sys

def part1(scan):
    position, depth = 0, 0
    for command in scan:
        direction, amount = command.strip().split()
        number = int(amount)
        if direction == "forward":
            position += number
        elif direction == "down":
            depth -= number
        elif direction == "up":
            depth += number
    return (position, depth)

def part2(scan):
    position, depth, aim = 0, 0, 0
    for command in scan:
        direction, amount = command.strip().split()
        number = int(amount)
        if direction == "forward":
            position += number
            depth += aim * number
        elif direction == "down":
            aim -= number
        elif direction == "up":
            aim += number
    return (position, depth)

def main(args = ()):
    arg = "day2.txt" if len(args) < 1 else args[0]
    with open(arg, "r") as scan:
        x,  y = part1(scan)
        print("Part 1:", abs(x*y))
        scan.seek(0)
        x,  y = part2(scan)
        print("Part 2:", abs(x*y))


if __name__ == "__main__":
    sys.exit(main(sys.argv[1:]))

File: day2.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>

using namespace std;

using Command = pair<string, int>;
using Commands = vector<Command>;

long part1(Commands& commands)
{
    int position = 0;
    int depth = 0;
    for (auto& command : commands) {
        if (command.first == "forward") {
            position = position + command.second;
        } else if (command.first == "down") {
            depth = depth + command.second;
        } else if (command.first == "up") {
            depth = depth - command.second;
        }
    }
    return position * depth;
}

long part2(Commands& commands)
{
    int position = 0;
    int aim = 0;
    long depth = 0;
    for (auto& command : commands) {
        if (command.first == "forward") {
            position = position + command.second;
            depth = depth + (aim * command.second);
        } else if (command.first == "down") {
            aim = aim + command.second;
        } else if (command.first == "up") {
            aim = aim - command.second;
        }
    }
    return position * depth;
}


int main(int argc, char* argv[])
{
    ifstream input;
    input.open(argc < 2 ? "day2.txt" : argv[1], ifstream::in);

    Commands commands;
    string direction;
    int distance;

    while (input.good()) {
        input >> direction >> distance;
        if (!input.good()) break; // account for an incomplete input, e.g. a blank line
        Command command(direction, distance);
        commands.push_back(command);
    }

    cout << "Part 1: " << part1(commands) << "\n";
    cout << "Part 2: " << part2(commands) << "\n";
}